rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+    // - copy the head block of data into that blob, using a stable memory read of a blob.
+    // - copy the head block of data from temp blob into new "final block" (logically still first) for region 0.
+    // - initialize the meta data for the region system in vacated initial block.
+
+    use crate::ic0_stable::nicer::{grow, read, write};
+
+    let header_len = meta_data::size::BLOCK_IN_BYTES as u32;
+
+    //let region0_pages = size() as u32;
+    let region0_pages = region_get_mem_size(mem) as u32;
+
+    let region0_blocks =
+        (region0_pages + (meta_data::size::PAGES_IN_BLOCK - 1)) / (meta_data::size::PAGES_IN_BLOCK);
+    assert!(region0_blocks > 0);
+
+    grow(
+        (meta_data::size::PAGES_IN_BLOCK + // <-- For new region manager
+	 /* Bump out region0 to nearest block boundary: */
+	 region0_blocks * meta_data::size::PAGES_IN_BLOCK
+            - region0_pages)
+            .into(),
+    );
+
+    // Temp for the head block, which we move to be physically last.
+    let header_val = crate::memory::alloc_blob(mem, crate::types::Bytes(header_len as u32));
+    let header_blob = header_val.as_blob_mut();
+    let header_bytes: &mut [u8] =
+        core::slice::from_raw_parts_mut(header_blob.payload_addr(), header_len as usize);
+
+    // Move it:
+    read(0, header_bytes); // Save first block as "head block".
+    write(
+        (region0_blocks * (meta_data::size::BLOCK_IN_BYTES as u32)).into(),
+        header_bytes,
+    );
+
+    // Zero-out the temp blob.
+    for i in 0..header_bytes.len() {
+        header_bytes[i] = 0;
+    }
+    write(0, header_bytes); // Zero out first block, for region manager meta data.
+
+    /* Initialize meta data as if there is only region 0. */
+    meta_data::total_allocated_blocks::set(region0_blocks.into());
+    meta_data::total_allocated_regions::set(1);
+
+    /*
+     * Initialize region0's table entries, for "recovery".
+     */
+
+    /* head block is physically last, but logically first. */
+    let head_block_id = region0_blocks - 1; /* invariant: region0_blocks is non-zero. */
+
+    meta_data::block_region_table::set(BlockId(head_block_id as u16), Some((RegionId(0), 0)));
+    meta_data::region_table::set(&RegionId(0), Some(RegionSizeInPages(region0_pages as u64)));
+
+    /* Any other blocks that follow head block are numbered [0,...,head_block_id) */
+    /* They're logical placements are [1,...,) -- one more than their new identity, as a number. */
+    for i in 0..head_block_id {
+        meta_data::block_region_table::set(
+            BlockId((i) as u16),
+            Some((RegionId(0), (i + 1) as u16)),
+        );
+    }
+    /* "Recover" the region data into a heap object. */
+    REGION_0 = region_recover(mem, &RegionId(0));
+
+    // Ensure that regions 2 through 15 are already reserved for
+    // future use by future Motoko compiler-RTS features.
+    region_reserve_id_span(mem, Some(RegionId(1)), RegionId(15));
```

Comment:
Maybe, a constant for the reserved region id number space (`15`) could be helpful, as it is used in each of the migration logics.